### ISSUE TITLE
Rule proposal: `prefer-module`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1067]
Should fail on all legacy commonjs stuff.

## Fail

```js
const foo = require('./foo');
```

```js
const foo = path.join(__dirname, './foo.js');
```

```js
module.exports = {};
```

```js
exports.foo = {};
```

## Pass

```js
export default foo;
```

```js
const foo = path.join(path.dirname(url.fileURLToPath(import.meta.url)), './foo.js');
```

**Comment by sindresorhus** [comment]
How about just `prefer-module` or `prefer-modules`? Alternatively, `prefer-js-module`.

**Comment by sindresorhus** [comment]
Should also fail on `__filename` and `'use strict';`.

**Comment by fisker** [comment]
Is `globalReturn` also commonjs only?

**Comment by fisker** [comment]
> `prefer-module`

👍

**Comment by sindresorhus** [comment]
> Is globalReturn also commonjs only?

Yes. That should be flagged too.

**Comment by lukeapage** [comment]
https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-commonjs.md

**Comment by sindresorhus** [comment]
@lukeapage That plugin is not actively maintained, the rule only covers some of what's mentioned here, and the plugin is slow as it parses all the imports and their graph.

**Comment by lukeapage** [comment]
it is fairly actively maintained and only rules that require traversing imports cause that slow behavior- just trying to help you.

**Comment by sindresorhus** [comment]
> it is fairly actively maintained

~17 real commits since 2018, a year since the last minor release, and 64 open PRs, is not what I would call actively maintained. Yes, it's maintained, but it seems its scope and maintenance requirements are larger than what's available. I have personal experience with this. I added the plugin to [`xo`](https://github.com/xojs/xo) years ago and it has been a continuous pain of bugginess.

> only rules that require traversing imports cause that slow behavior

Ah. I didn't look at the code. I just have experience with many of the rules being slow.

> just trying to help you.

And I appreciate that. I'm just arguing why this rule is still worth doing.

**Comment by sindresorhus** [comment]
Just want to point out that:

```js
const foo = path.join(path.dirname(url.fileURLToPath(import.meta.url)), './foo.js');
```

could be simplified to:

```js
const foo = fileURLToPath(new URL('foo.js', import.meta.url));
```

and in many cases you could just do this as `fs` methods support `URL`:

```js
const foo = new URL('foo.js', import.meta.url);
```

We should note this in the docs.

**Comment by sindresorhus** [comment]
We should also documen `__dirname` / `__filename` replacements:

```js
import {fileURLToPath} from 'url';
import path from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(fileURLToPath(import.meta.url));
```

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by fisker** [comment]
I already get started, I was thinking add suggestions for these replacements for `__dirname` and `__filename`, but user need import `url` and `path` module themself.

**Comment by sindresorhus** [comment]
> I was thinking add suggestions for these replacements for __dirname and __filename, but user need import url and path module themself.

Why not an auto-fix? And just add the imports?

**Comment by sindresorhus** [comment]
We should also ban the whole `module` and `require` objects. We don't want users to do `module.parent` or `require.resolve`.

For the one thing they need `module` for, it's better to explicitly import it. We should allow this one:

```js
import {createRequire} from 'module';
```

**Comment by sindresorhus** [comment]
We should also enforce using `.js` extension for relative imports, this also applies to TypeScript (https://github.com/typescript-eslint/typescript-eslint/issues/3288).

**Comment by fisker** [comment]
I think it's unrelated to this rule.

**Comment by sindresorhus** [comment]
@fisker The point of this rule is to enforce ESM, correct? ESM requires a `.js` extension for relative imports. I don't understand how it's unrelated.

**Comment by fisker** [comment]
What if it mean to

```js
import foo from './foo.vue'
```

and

```js
import foo from './foo'
```

just work fine.

**Comment by sindresorhus** [comment]
Then it's not native ESM. This rule should be about native ESM. Not Babel/Webpack's version. Native ESM requires extension no matter what. Vue files will eventually be supportable natively through loader hooks and the imports will then require the `.vue` extension. It's better to just switch now.

**Comment by fisker** [comment]
If we want do it, we check import/export source, if it's not endWith `.js` add `.js`? Or should it only check `require`? Why not a separate rule enforce it?

**Comment by sindresorhus** [comment]
It could be a separate rule, sure. I just thought it would fit this rule.

**Comment by sindresorhus** [comment]
> If we want do it, we check import/export source, if it's not endWith .js add .js?

Yes, but only for relative imports.

**Comment by fisker** [comment]
> I just thought it would fit this rule.

If add to this rule, it make sense to fix the require source adding `.js`, but doesn't make sense to check `import`/`export` in this rule, let's do it in another rule, then we can use the same logic to update the `require` fix part in this rule.

**Comment by sindresorhus** [comment]
@fisker Done: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1194
